### ISSUE TITLE
[data grid] Reduce filter debounce from 500ms to 50ms

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-x#9657]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

We have a 500ms input debounce that feels like the biggest chunk of time where end-users are waiting for the data grid to return results. I believe that even with an instant filtering, this denounce makes it impossible for the search input to feel delightful.

https://github.com/mui/mui-x/blob/df534be519b791031294577c15eb9176be399279/packages/grid/x-data-grid/src/components/toolbar/GridToolbarQuickFilter.tsx#L74

https://github.com/mui/mui-x/blob/82ab9c02d7381b37376ac25323d69685b43f89d9/packages/grid/x-data-grid/src/components/panel/filterPanel/GridFilterInputValue.tsx#L8

I suspect that we could fix it like this:

- lower the debounce to 50ms, we would save 450ms! (0ms would probably be fine too), for comparison, the description of #9120 talks about a win of up to 300ms.
- turn the search to be performed in chunks, to yield to the main thread every now and then
- start the search as soon as the end of the 50ms debounce but if a new search is input before the end of the previous one, **interrupt it** and restart from scratch.

### Examples 🌈

https://mui.com/x/react-data-grid/filtering/quick-filter/

**Comment by romgrk** [comment]
Linked #9157

**Comment by oliviertassinari** [comment]
A quick performance benchmark to illustrate the problem:

### MUI X

100,000 rows from https://mui.com/x/react-data-grid/#commercial-version, using a modified version of the demo to add

```jsx
        slots={{ toolbar: GridToolbar }}
        slotProps={{
          toolbar: {
            showQuickFilter: true,
          },
        }}
```

<img width="1422" alt="Screenshot 2023-07-13 at 00 08 10" src="https://github.com/mui/mui-x/assets/3165635/018dc8a5-7cf2-44ab-9cf1-f82ad4b7729f">

From the last key down to the first frame, we wait 1,430ms.

With the work on #9120 and related to it, I imagine we could move from 894ms spent on JS, down to 194ms, we would then get the wait from 1430ms to 730ms. By reducing the debounce, we could save another 450ms, going down to 280ms.

### AG Grid

100,000 rows from https://www.ag-grid.com/example/

<img width="1426" alt="Screenshot 2023-07-13 at 00 20 00" src="https://github.com/mui/mui-x/assets/3165635/14f012d3-30ea-4e36-a074-6502f95ff075">

From the last key down to the first frame, we wait 277ms. It feels like they don't have any debounce at all, but it's doesn't feel great, the input update is lagging, ewww 🙈

**Comment by ithrforu** [comment]
What if replace debounce with `useDefferedValue`? It helps us a lot with "heavy" inputs. Maybe it's similar situation.

**Comment by oliviertassinari** [comment]
@ithrforu `useDefferedValue` would solve AG Grid's problem (the input text that lags behind). In our case, I don't know, we could also consider a throttle, like it feels Netflix https://www.youtube.com/watch?v=FAZJsxcykPs or even our Algolia search on the docs is doing (though, they can run queries in parallel in the server, we can't really client side).

**Comment by oliviertassinari** [comment]
I'm reopening. We had a step forward in https://github.com/mui/mui-x/pull/9712, but 

- the input lags now, similar to https://github.com/mui/mui-x/issues/9657#issuecomment-1640080978. This doesn't feel responsive.
- we can't start searches without stopping them in the middle. We could save a lot of work with large tables.
- the debounce is at 150ms. It would likely be better at a lower value, like 50ms or lower, saving >100ms.

**Comment by romgrk** [comment]
1. Wdym by input lag? I didn't notice it.
2. It's going to require more work on the filtering, we need to decouple things a bit before we can get there.
3. We set it at 150ms because of 2, until we solve that it's preferable to have it at a speed where it's less likely to trigger between typing 2 characters.

**Comment by oliviertassinari** [comment]
> Wdym by input lag? I didn't notice it.

1. Two demos: see how "q" takes a long time to be updated in the `<input>`

https://github.com/mui/mui-x/assets/3165635/22f91512-eb00-4030-9eed-ab81e6a0c104

and see how select all + delete doesn't work:

https://github.com/mui/mui-x/assets/3165635/a4ab2021-a683-4d56-822d-e3db136b8e9c

**Comment by romgrk** [comment]
Yes so it feels like all of those issues are caused by the filter being triggered in-between the keypresses. The immediate solution to that would be to... increase the filter debounce again :sweat_smile: 

With the 150ms delay, the UX for grids with small datasets is better, but for large datasets it's worst. I have mentioned somewhere else that an adaptive debouce delay could be interesting (small dataset = small delay, large dataset = large delay).

And of course the ideal solution would be interruptable filtering, but that requires quite a bit of work and it's not on the roadmap at the moment afaik. We may want to discuss how to prioritize that.

**Comment by romgrk** [comment]
It doesn't feel like the same issue, it's not a logical bug, it's caused by the CPU doing filtering while the user is typing.

**Comment by oliviertassinari** [comment]
Coming from https://twitter.com/addyosmani/status/1767540608779083972, INP is now part of SEO ranking

Testing https://next.mui.com/x/react-data-grid/filtering/quick-filter/ using https://web.dev/articles/debug-cwvs-with-web-vitals-extension (Chrome extension) illustrates the problems well:

https://github.com/mui/mui-x/assets/3165635/bd9b4940-ebad-4fdf-affb-826ec2eb8fd1

This video is perfect https://youtu.be/KZ1kxzsJZ5g?si=DmhtZ8DZ0rbxs35-&t=223 and bring back the solution raised before. We need to 

1. We are not using `useTransition` to update the input
2. We are not filtering the row in chunks

**Comment by oliviertassinari** [comment]
One day, we could use `scheduler.yield()` https://web.dev/articles/top-cwv#in-long-tasks instead of:

https://github.com/mui/mui-x/blob/9b0e411eed1fb7621fba4c3fcb8b8bb692e2cf5b/packages/x-data-grid-generator/src/hooks/useDemoData.ts#L68

that has a "background" task priority inherited from requestIdleCallback() but we want "user-visible", no? https://developer.mozilla.org/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities

**Comment by oliviertassinari** [comment]
AG Grid made a step froward with https://github.com/mui/mui-x/issues/9657#issuecomment-1633283607. Proof https://youtu.be/ruSdTd-aATQ?si=-lsoYovFGj-Ct0NR&t=646
